pkg/backup/backup.go

```diff
@@ -174,94 +175,188 @@ func (b *Backup) Run() {
 	}
 }
 
+// saveSnap saves the latest snapshot if latest revision is greater than the given lastSnapRev.
 func (b *Backup) saveSnap(lastSnapRev int64) (int64, error) {
-	podList, err := b.kclient.Core().Pods(b.namespace).List(k8sutil.ClusterListOpt(b.clusterName))
+	start := time.Now()
+
+	snap, rev, err := GetSnap(b.kclient, b.etcdTLSConfig, b.namespace, b.clusterName)
 	if err != nil {
-		return lastSnapRev, err
+		return lastSnapRev, fmt.Errorf("failed to get snapshot: (%v)", err)
 	}
+	defer snap.Close()
 
-	var pods []*v1.Pod
-	for i := range podList.Items {
-		pod := &podList.Items[i]
-		if pod.Status.Phase == v1.PodRunning {
-			pods = append(pods, pod)
-		}
+	if rev <= lastSnapRev {
+		logrus.Info("skipped creating new backup: no change since last time")
+		return lastSnapRev, nil
 	}
 
-	if len(pods) == 0 {
-		msg := "no running etcd pods found"
-		logrus.Warning(msg)
-		return lastSnapRev, fmt.Errorf(msg)
-	}
-	member, rev := getMemberWithMaxRev(pods, b.etcdTLSConfig, b.selfHosted)
-	if member == nil {
-		logrus.Warning("no reachable member")
-		return lastSnapRev, fmt.Errorf("no reachable member")
+	etcdVersion, err := snap.Version()
+	if err != nil {
+		return lastSnapRev, fmt.Errorf("failed to get etcd version: (%v)", err)
 	}
 
-	if rev <= lastSnapRev {
-		logrus.Info("skipped creating new backup: no change since last time")
-		return lastSnapRev, nil
+	n, err := WriteSnap(etcdVersion, rev, b.be, snap)
+	if err != nil {
+		return lastSnapRev, fmt.Errorf("failed to write snapshot to backend: (%v)", err)
 	}
 
-	log.Printf("saving backup for cluster (%s)", b.clusterName)
-	if err := b.writeSnap(member, rev); err != nil {
-		err = fmt.Errorf("write snapshot failed: %v", err)
-		return lastSnapRev, err
+	bs := backupapi.BackupStatus{
+		CreationTime:     time.Now().Format(time.RFC3339),
+		Size:             util.ToMB(n),
+		Version:          etcdVersion,
+		Revision:         rev,
+		TimeTookInSecond: int(time.Since(start).Seconds() + 1),
 	}
+	b.recentBackupsStatus = append(b.recentBackupsStatus, bs)
+	if len(b.recentBackupsStatus) > maxRecentBackupStatusCount {
+		b.recentBackupsStatus = b.recentBackupsStatus[1:]
+	}
+
 	return rev, nil
 }
 
-func (b *Backup) writeSnap(m *etcdutil.Member, rev int64) error {
-	start := time.Now()
+// WriteSnap writes the snapshot to the backend.
+func WriteSnap(version string, rev int64, be backend.Backend, snap *Snap) (int64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
+	defer cancel()
```

Comment:
i used defer here so that `cancel` is called when both `rc, err := snap.Snapshot(ctx)` error out and `n, err := be.Save(version, rev, rc)` error out.